app/maptools/measurementmaptool.h

```diff
@@ -0,0 +1,132 @@
+    Q_PROPERTY( double perimeter READ perimeter WRITE setPerimeter NOTIFY perimeterChanged )
+    Q_PROPERTY( double area READ area WRITE setArea NOTIFY areaChanged )
+
+    Q_PROPERTY( bool canUndo READ canUndo WRITE setCanUndo NOTIFY canUndoChanged )
+    Q_PROPERTY( bool canCloseShape READ canCloseShape WRITE setCanCloseShape NOTIFY canCloseShapeChanged )
+    Q_PROPERTY( bool closeShapeDone READ closeShapeDone WRITE setCloseShapeDone NOTIFY closeShapeDoneChanged )
+
+  public:
+    explicit MeasurementMapTool( QObject *parent = nullptr );
+    virtual ~MeasurementMapTool() override;
+
+    /**
+     * Adds point to the end of the recorded geometry; updates recordedGeometry afterwards
+     * Passed point needs to be in map CRS
+     */
+    Q_INVOKABLE void addPoint();
+
+    /**
+     *  Removes last point from recorded geometry if there is at least one point
+     *  Updates recordedGeometry afterwards
+     */
+    Q_INVOKABLE void removePoint();
+
+    /**
+     * If there are at least 3 points, forms a polygon from recorded points so far.
+     * Updates recorded geometry and calculates area and perimeter of formed polygon.
+    */
+    Q_INVOKABLE void closeShape();
+
+    /**
+     * Repeats measurement process.
+     * Clears all recorded points and rebuilds the geometry.
+     */
+    Q_INVOKABLE void reset();
+
+    // Getter and Setters
+    double lengthWithGuideline() const;
+    void setLengthWithGuideline( const double &length );
+
+    double perimeter() const;
+    void setPerimeter( const double &perimeter );
+
+    double area() const;
+    void setArea( const double &area );
+
+    QPointF crosshairPoint() const;
+    void setCrosshairPoint( const QPointF &point );
+
+    bool canUndo() const;
+    void setCanUndo( bool newCanUndo );
+
+    bool canCloseShape() const;
+    void setCanCloseShape( bool newCanCloseShape );
+
+    bool closeShapeDone() const;
+    void setCloseShapeDone( bool newCloseShapeDone );
+
+    const QgsGeometry &recordedGeometry() const;
+    void setRecordedGeometry( const QgsGeometry &newRecordedGeometry );
+
+    QgsGeometry existingVertices() const;
+    void setExistingVertices( const QgsGeometry &vertices );
+
+  signals:
+    void lengthWithGuidelineChanged( const double &lengthWithGuideline );
+    void perimeterChanged( const double &perimeter );
+    void areaChanged( const double &area );
+    void canUndoChanged( bool canUndo );
+    void canCloseShapeChanged( bool canUndo );
+    void closeShapeDoneChanged( bool canUndo );
+    void recordedGeometryChanged( const QgsGeometry &recordedGeometry );
+    void existingVerticesChanged( const QgsGeometry &vertices );
+    void crosshairPointChanged( const QPointF &crosshairPoint );
+
+  protected:
+    void rebuildGeometry();
+    void checkCanCloseShape();
+
+  public slots:
+    void updateDistance();
+
+  private slots:
+    void onMapSettingsChanged( InputMapSettings *newMapSettings );
```

Comment:
You _should_ be able to declare this as `override`, I believe even without declaring it as a slot. Then work with it as a derived method from base class.

```suggestion
    void onMapSettingsChanged( InputMapSettings *newMapSettings ) override;
```

Reply:
Wrong suggestion! We need to override the setter, not the signal 🤦🏻